fairscale/utils/parallel.py

```diff
@@ -104,12 +111,19 @@ def get_process_group_cached(
     if not dist.is_initialized():
         # Likely caused by initiating a dummy pg for unit test, skip checking.
         if name == ProcessGroupName.reduce_scatter and "pytest" in sys.modules:
+            # FIXME: consider dummy having an explicit marker or base class,
+            # and having an explicit `is_dummy_process_group(pg)` test,
+            # rather than assuming that malformed pgs are dummy pgs,
+            # and doing the test differently in each of these places.
             return None
         else:
             raise RuntimeError("torch.distributed is not yet initialized but process group is requested.")
 
+    # FIXME: it is _very_ odd to hang this cache off a function attribute,
```

Comment:
True. This is like a hack. We have different needs in deals with PGs. They impact GPU memory as well as overlapping of compute & communication. They also impact usability and flexibility in terms of letting users control PG or letting us do it. The caching method here is by no means the best possible method.